src/posix/platform/netif.cpp

```diff
@@ -384,6 +427,122 @@ static void processNetifEvent(otInstance *aInstance)
+
+    VerifyOrExit(setsockopt(sMLDMonitorFd, IPPROTO_IPV6, IPV6_JOIN_GROUP, &mreq6, sizeof(mreq6)) == 0,
+                 err = OT_ERROR_FAILED);
+    VerifyOrExit(setsockopt(sMLDMonitorFd, SOL_SOCKET, SO_BINDTODEVICE, sTunName,
+                            static_cast<socklen_t>(strnlen(sTunName, IFNAMSIZ))) == 0,
+                 err = OT_ERROR_FAILED);
+
+exit:
+    if (err != OT_ERROR_NONE)
+    {
+        if (sMLDMonitorFd >= 0)
+        {
+            close(sMLDMonitorFd);
+        }
+        DieNowWithMessage("Failed to initialize MLD listener", OT_EXIT_FAILURE);
+    }
+}
+
+static bool isMulticastAddressFiltered(const struct in6_addr &aAddress)
+{
+    bool found = false;
+
+    for (size_t i = 0; i < sizeof(kFilterMulticastAddresses) / sizeof(kFilterMulticastAddresses[0]); i++)
+    {
+        if (memcmp(&aAddress, &kFilterMulticastAddresses[i].mFields.m8, sizeof(aAddress)) == 0)
+        {
+            found = true;
+            break;
+        }
+    }
+
+    return found;
+}
+
+static void processMLDEvent(otInstance *aInstance)
+{
+    const size_t        kMaxMLDEvent = 8192;
+    uint8_t             buffer[kMaxMLDEvent];
+    ssize_t             bufferLen = -1;
+    struct sockaddr_in6 srcAddr;
+    socklen_t           addrLen;
+    bool                fromSelf = false;
+    MLDv2Header *       hdr      = reinterpret_cast<MLDv2Header *>(buffer);
+    size_t              offset;
+    uint8_t             type;
+    struct ifaddrs *    ifAddrs = NULL;
+
+    bufferLen = recvfrom(sMLDMonitorFd, buffer, sizeof(buffer), 0, reinterpret_cast<sockaddr *>(&srcAddr), &addrLen);
+    VerifyOrExit(bufferLen > 0);
+
+    type = buffer[0];
+    VerifyOrExit(type == kICMPv6MLDv2Type && bufferLen >= static_cast<ssize_t>(sizeof(MLDv2Header)));
+
+    // Check whether it is sent by self
+    VerifyOrExit(getifaddrs(&ifAddrs) == 0);
+    for (struct ifaddrs *ifAddr = ifAddrs; ifAddr != NULL; ifAddr = ifAddr->ifa_next)
+    {
+        if (ifAddr->ifa_addr != NULL && ifAddr->ifa_addr->sa_family == AF_INET6 &&
+            strncmp(sTunName, ifAddr->ifa_name, IFNAMSIZ) == 0)
+        {
+            struct sockaddr_in6 *addr6 = reinterpret_cast<struct sockaddr_in6 *>(ifAddr->ifa_addr);
+
+            if (memcmp(&addr6->sin6_addr, &srcAddr.sin6_addr, sizeof(in6_addr)) == 0)
+            {
+                fromSelf = true;
+                break;
+            }
+        }
+    }
+    VerifyOrExit(fromSelf);
+
+    hdr    = reinterpret_cast<MLDv2Header *>(buffer);
+    offset = sizeof(MLDv2Header);
+
+    for (size_t i = 0; i < hdr->mNumRecords && offset < static_cast<size_t>(bufferLen); i++)
+    {
+        if (static_cast<size_t>(bufferLen) >= (sizeof(MLDv2Record) + offset))
+        {
+            MLDv2Record *record = reinterpret_cast<MLDv2Record *>(&buffer[offset]);
+
+            if (!isMulticastAddressFiltered(record->mMulticastAddress))
```

Comment:
FWIW, in the mac OS code I included in #4664, I do catch the failures and AT LEAST log them.  I really hate that these things fail silently.  I don't mind if the system tries to keep running in a "not completely working" state, but we should at least complain about it at runtime.

Reply:
I agree that we should log any errors returned from `otIp6SubscribeMulticastAddress()` and `otIp6UnsubscribeMulticastAddress()`.

Reply:
Also FWIW, I need to develop a better understanding of how OpenThread's logging mechanisms work.  My code just spits stuff to stderr using fprintf -- since this code is on POSIX, that turns out to be OK.  But I'm suspicious there's a way to use some ot logging APIs and get the same results (i.e., errors and interesting results get spit to stderr).  Using that seems "more correct" for this codebase.  And I'm hopeful that I wouldn't have to provide my own "IPv6-address-to-string" function if I was using the preferred way.  But back when I started on this effort (almost a year ago), the ot logging interfaces were more challenging and less full-featured that I could deal with.

Reply:
I filter these addresses to prevent subscribing ff02::02 when thread is enabled. I'm not quit sure about  the current codes behavior under such circumstance but I decided to filter them to be safe.

Reply:
Logging the error from the `otIp6SubscribeMulticastAddress()` and `otIp6UnsubscribeMulticastAddress()` sounds good to me, but the question for me is if we want to do the filtering of addresses in this module or not.

I would say, unless we change the code in this module to `assert` on an error,  it may be better not to filter/exclude any multicast address and log the error. Particularly since the change in https://github.com/openthread/openthread/pull/4693 makes the returned error more appropriate when address is already present on the list. We then treat `OT_ERROR_ALREADY` and `OT_ERROR_NONE` as ok/success cases.

About `ff02::02` the OT API itself should not allow an external subscription for it and I expect would return the appropriate error code (OT_ERROR_ALREADY if we are FTD (router/reed), and OT_ERROR_INVALID_ARGS if we are MTD). 

Thoughts?

Reply:
If OT will do the filtering than that looks fine to me.